### ISSUE TITLE
funannotate Bioconda recipe

### ISSUE THREAD

**Issue by nextgenusfs** [nextgenusfs/funannotate#194]
I would like to make a bioconda package for funannotate -- as it is still difficult to install. Would be great if a conda expert would be willing to help out. So far there are some dependencies still missing would need to get done first.

1) Evidence Modeler is not in bioconda
2) Trinity and PASA are currently Linux only (not available for macOS) -- seems to be a compiler incompatibility?
3) Augustus via bioconda is not fully functional on macOS -- I use a slightly modified version of [augustus v3.2.1](https://github.com/nextgenusfs/augustus)
4) funannotate code is currently python2 only (needs to be migrated to py2/3 compatible) -- note this is perhaps not required for a recipe but would help with future compatibility. 

Any help/guidance would be appreciated!

**Comment by hyphaltip** [comment]
We started something for FGMP and the problems with Augustus 3.3 on bioconda also a problem.

Maybe we can tweet about this some of the bioconda devs are perhaps able to contribute suggestions.

Jason Stajich, PhD
jasonstajich.phd@gmail.com

**Comment by kastman** [comment]
Hi @nextgenusfs @hyphaltip ,

I'm just discovering `funannotate` now and am impressed; thanks for the work! 

I helped commit the OSX `augustus` bioconda build (about 5 months ago) and wasn't aware it was breaking - do you have an issue to point me to for more of a failure log? 

I see in reading the notes from your [custom macOS augustus build](https://github.com/nextgenusfs/augustus) that it appears to be a compiler compatibility problem with bamtools, causing the proteinprofile BUSCO search to fail; is that it? Maybe we could add a test directly to the augustus recipe to catch that if handling both bamtools and augustus installs with conda doesn't catch it.

As for trinity and pasa, I'm not sure what's going on. Bioconda has been in the process of rebuilding **all** their recipes for the last 3 weeks to use the newer conda build system, and things are breaking a bit - for example there is a [pasa recipe](https://github.com/bioconda/bioconda-recipes/tree/master/recipes/pasa) (that appears to support OSX) but it doesn't show up on the [bioconda search page](https://bioconda.github.io/recipes.html). Might be worth focusing on other development for a few weeks to let the dust settle on this one, and then swing back around to it.

I'm happy to help out, as a widespread euk annotation pipeline is definitely needed. Thanks!

**Comment by kastman** [comment]
PS - As I'm literally brand new to `funannotate`, I'm currently stepping through the tutorial and will use the bioconda bamtools and augustus installs; most likely I'll find the error on my own. :) But just in case I don't, send a link if you have it. Thanks,

**Comment by kastman** [comment]
I verified your error with bioconda augustus `3.2.3` and `3.3`, and read through what you've tried already in #3. Do you know what about the newer gcc does the fix? If we knew what was changed, we could possibly add a patch option to the older gcc. Compiling on different versions of gcc is one of the few limitations of conda-forge/bioconda, which is quite strict (to make sure everything is compatible), so just bumping the gcc version isn't easy / possible, though I think the recent rebuild/overhaul is doing just that. I somehowd doubt that adding 3.2.1 to bioconda would help if it's also still compiled with the older gcc.

I know you haven't had much luck getting feedback, but maybe @mariostanke would be able to weigh in (not sure if he's getting github notifications though)?

Sorry to overtake this ticket - maybe we should move this conversation back to #3?

**Comment by nextgenusfs** [comment]
Yes would certainly be best to be fixed upstream.  Its possible that adding `3.2.1` might actually help, as the `--proteinprofile` part of Augustus seems to compile properly on OSX using a newer GCC (I've tried gcc-5 through gcc-7 I think). The changes that I made in https://github.com/nextgenusfs/augustus were just in the Makefiles as it is largely fixing the bamtools compilation problems with `filterbam` and `bam2hints` -- quite likely there is/was a different/better way to fix those errors. 

Is there a way to test a local bioconda build with `3.2.1` using the existing patches/recipes for `3.2.3` and `3.3` on OSX?  I did also see there is a new release http://bioinf.uni-greifswald.de/augustus/binaries/, have not tested it yet.

**Comment by kastman** [comment]
Augustus was on the bioconda blacklist because the author moved the tarball into an "old" directory and broke the link -- I'm fixing that and will test 3.2.1 as well as update for the recent 3.3.1 release. 

You can test a local build by forking [bioconda/bioconda-recipes](https://github.com/bioconda/bioconda-recipes), updating the version and sha256 hash in `recipes/augustus`, and running `circleci build`. I'm going to give that a shot and let you know. I suspect it's a compiler problem (still running 4.8) so this likely won't work, but worth a shot?

**Comment by nextgenusfs** [comment]
So after you build with `circleci build` can you then test those packages?  I've got one of the other software packages, `AMPtk` onto bioconda, but still very much a noob when it comes to the details of the build and the testing. But either way can update this thread with results of bioconda builds for `3.2.1` and `3.3.1` would be helpful.

**Comment by kastman** [comment]
If you want to test locally, I usually build using `conda build` directly, and then `conda install --use-local`, which uses the new package you just built (using `circleci` actually builds in an image which is then tossed). 

I'm sure the new compiler changes in conda-forge/bioconda will be useful, but there are still some kinds: now it seems there's a problem finding some of the perl modules (yaml). I'll keep you posted, but I'm still working on getting it off the blacklist.

I also noticed the [specialized 3.2.1 PR](https://github.com/bioconda/bioconda-recipes/pull/8243) from @camillescott - pulling from Jon's fixed tarball isn't the way bioconda is intended to be used, but may not be the worst idea? Ideally I'd pull in your specific patches and apply them to the upstream tarball.

Regardless, we have to clear it off the blacklist first. I'll let you know how that goes or you can follow along with the [PR](https://github.com/bioconda/bioconda-recipes/pull/9853). 

Thanks!  

(also adding @lizlandis to this conversation as well so she can see the progress).

**Comment by hmontenegro** [comment]
How is this moving along? 

I think it will be difficult to have a bioconda recipe for funannotate and all its dependencies: when I tried to install as many dependencies as possible from conda, there were some packages which conflicted with each other. For example, Augustus installed, but then did not run due to wrong Boost version. Depending on the combination of packages, a different Numpy would be installed, and then loading Numpy segfaulted. And some other similar small problems.

Maybe a better solution would be to create a funannotate Anaconda channel, where developers / contributors would have more control over versions installed, and over applied patches. 

I just started learning how to build conda recipes, and I don't have access to Macs, only Linux. That said, I could help creating / testing a conda recipe.

**Comment by nextgenusfs** [comment]
Difficult = yes.  However, that is the whole reason to get a conda recipe together - to avoid the dependency nightmare.  I don' t have much experience with conda and no experience with setting up "channels".  Shouldn't we be okay if we specify package version numbers?

I know when bioconda updated recently there are a lot of packages that have problems and need to be rebuilt. 

We could forge ahead with a linux only recipe -- Augustus on Mac is still not solved (still issues with compilation I think).  I wrote an EvidenceModeler recipe here https://github.com/bioconda/bioconda-recipes/pull/10389 but hasn't been merged yet.

**Comment by kastman** [comment]
Sorry to have dropped this - I didn't make any progress with augustus myself, but it was taken off the blacklist and fixed in August in [this PR](https://github.com/bioconda/bioconda-recipes/pull/10364). Looking now to see if a backported 3.2.1 will fix the compilation.

Certainly a linux-only recipe is better than nothing and the mac problems shouldn't hold it up, though that's where most of my time is spent and where I'm most motivated to keep up the good fight. :)

**Comment by nextgenusfs** [comment]
Hi @kastman, I was just running the Augustus v3.3 on Linux and noted that the bioconda install "breaks" `funannotate`.  In `funannotate`, I use the `$AUGUSTUS_CONFIG_PATH` to get the augustus "base directory" which then I use to get the location of the `scripts` folder, so it calls the necessary accessory scripts by full path -- thus alleviating the user from also putting the scripts folder in their PATH.  I noticed that the bioconda install copies these over to `/bin` so they are in $PATH. But then these scripts are not found by `funannotate` the pipeline crashes.   The workaround I used post bioconda `augustus` install is then to symlink a `/scripts/` to `/bin/` directory -- so that the `augustus` directory tree is intact.  I totally understand why scripts would be copied over to `bin`, however, that isn't how most manual `augustus` install "look". For other packages that also use a lot of "accessory" scripts, i.e. `pasa` and `trinity` the bioconda recipes put the entire install folder into `/conda/opt` which then keeps the directory tree intact -- that would be preferable to me. As if I change the way funannotate looks for these scripts -- it will likely break those pipelines where `augustus` is manually installed.

**Comment by kastman** [comment]
Hi @nextgenusfs -- yep, copying the install to `/conda/opt` sounds like a reasonable step, and it sounds like there's precedent; I'll take a deeper look when I get a second. Is it possible to adjust the `$AUGUSTUS_CONFIG_PATH` to point to the tree correctly too?

I tried to rebuild 3.2.1 now that augustus is off the blacklist over the weekend, but the patch didn't apply properly and I ran out of time. However, I noticed that the big conda update / rebuild is using GCC7, which might fix the compilation problem that 3.2.3 had if it gets rebuilt. 

I'll let you know as I figure out more, but that certainly sounds reasonable.

**Comment by nextgenusfs** [comment]
Yes I think can set the ENV variables like are done in pasa/trinity, i.e. in the `build.sh` via https://github.com/bioconda/bioconda-recipes/blob/master/recipes/pasa/build.sh
```
readonly PASAHOME=${PREFIX}/opt/${PKG_NAME}-${PKG_VERSION}

mkdir -p ${PASAHOME}
cp -Rp bin Launch_PASA_pipeline.pl misc_utilities pasa_conf PasaWeb PasaWeb.conf PerlLib PyLib run_PasaWeb.pl SAMPLE_HOOKS schema scripts ${PASAHOME}

mkdir -p ${PREFIX}/etc/conda/activate.d/
echo "export PASAHOME=${PASAHOME}" > ${PREFIX}/etc/conda/activate.d/${PKG_NAME}-${PKG_VERSION}.sh

mkdir -p ${PREFIX}/etc/conda/deactivate.d/
echo "unset PASAHOME" > ${PREFIX}/etc/conda/deactivate.d/${PKG_NAME}-${PKG_VERSION}.sh
```
So this basically copies over the PASA directory tree to /opt/pasa-2.3.3. So for `augustus` would be similar I think, do the same install as before and copy over the necessary files to `/opt` then I think just need to link `augustus` to `/bin` and then entire folder structure would be same as install with executable symlinked into conda $PATH. 

Here's a quick stab at it (untested):
```
#!/bin/bash

set -x -e

export INCLUDE_PATH="${PREFIX}/include"
export LIBRARY_PATH="${PREFIX}/lib"
export LD_LIBRARY_PATH="${PREFIX}/lib"
export BOOST_INCLUDE_DIR=${PREFIX}/include
export BOOST_LIBRARY_DIR=${PREFIX}/lib

#export CXXFLAGS=" -std=c++11 -stdlib=libstdc++ -stdlib=libc++ -DUSE_BOOST -I${BOOST_INCLUDE_DIR} -L${BOOST_LIBRARY_DIR}"
export CXXFLAGS=" -std=c++11  -DUSE_BOOST -I${BOOST_INCLUDE_DIR} -L${BOOST_LIBRARY_DIR}"
export LDFLAGS="-L${BOOST_LIBRARY_DIR}"

#setup directories
AUG_HOME=$PREFIX/opt/augustus-$PKG_VERSION
mkdir -p $PREFIX/bin
mkdir -p $AUG_HOME

## Make the software

sed -i.bak -e 's/^CC *=/CXX=/' -e 's/\$(CC)/$(CXX)/g' auxprogs/homGeneMapping/src/Makefile
sed -i.bak -e 's/^CC *=/CXX=/' -e 's/\$(CC)/$(CXX)/g' auxprogs/joingenes/Makefile
# TODO: don't set CC/CXX here when switching to newer compilers
CC=gcc
CXX=g++
if [ "$(uname)" == Darwin ] ; then
  # SQLITE disabled due to compile issue, see: https://svn.boost.org/trac10/ticket/13501
  make CC="${CC}" CXX="${CXX}" COMPGENPRED=true
else
  make CC="${CC}" CXX="${CXX}" COMPGENPRED=true SQLITE=true
fi

## Build Perl

mkdir perl-build
find scripts -name "*.pl" | xargs -I {} mv {} perl-build
cd perl-build
cp ${RECIPE_DIR}/Build.PL ./
perl ./Build.PL
perl ./Build manifest
perl ./Build install --installdirs site

cd ..

## End build perl

cp -Rp scripts config ${AUG_HOME}
mv bin/* $PREFIX/bin/ 

#Add some options to activate

mkdir -p $PREFIX/etc/conda/activate.d/
echo "export AUGUSTUS_CONFIG_PATH=${AUG_HOME}/config/" > $PREFIX/etc/conda/activate.d/augustus-confdir.sh
chmod a+x $PREFIX/etc/conda/activate.d/augustus-confdir.sh

mkdir -p $PREFIX/etc/conda/deactivate.d/
echo "unset AUGUSTUS_CONFIG_PATH" > $PREFIX/etc/conda/deactivate.d/augustus-confdir.sh
chmod a+x $PREFIX/etc/conda/deactivate.d/augustus-confdir.sh

chmod u+rwx $PREFIX/bin/*
```
